### PR TITLE
DAOS-11406 rdb: Use VOS local TXs

### PR DESCRIPTION
**Pull request by liw**
Every RDB (update) TX invokes multiple VOS writes, each of which
performs a WAL commit. This patch reduces those WAL commits to one per
RDB TX by utilizing the new VOS local TXs to group all VOS writes of one
RDB TX into one VOS local TX.

  - An rdb_vos_tx_t handle is introduced to hide the vos/dtx details
    from the rest of rdb.

  - Some rdb_lc_discard calls are no longer necessary, for the VOS local
    TX mechanism will discard uncommited writes more efficiently.

  - Some "nvops" ("number of VOS TX operations") functions are added to
    provide dtx_begin with an upper bound of the number of VOS TX
    operations that will be invoked using a VOS TX handle. Hopefully
    this will become unnecessary in the future.

  - The rdb_tx_hdr decoding in rdb_tx_apply should use the actual length
    of the data, rather than the expected length.

  - Two places related to rdb_tx_hdr need to be refactored a bit in
    order to avoid compiler warnings along the line of "hdr may be used
    uninitialized". It's unclear why the compiler warnings suddenly come
    up.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by liw** [comment]
@jolivier23, @kccain, just in case you are interested, I requested _draft_ reviews to make this PR easier for you to find.

**Comment by liw** [comment]
@kccain, @jolivier23, could we do an initial round of review before the VOS part is ready? I have done some basic testing with one of @janekmi's WIP commit, but the full CI testing cannot start before the VOS part is ready.

File: src/rdb/rdb_util.c
Comment (liw) [inline src/rdb/rdb_util.c:230]: [Note] This temporary compilation enables the current PR to build without the VOS changes. It will be removed once the VOS changes land.

File: src/rdb/rdb_raft.c
Comment (kccain) [inline src/rdb/rdb_raft.c:1119]: Sorry if it is evident / should be known already. This is the last file of the set I'm looking at and may be losing steam. :)

For rdb_raft_log_offer_single() this value (2) corresponds to the rdb_lc_update() call and the rdb_mc_update() call in that function? Or, does it refer to the (potential) n=2 keys/values fed into rdb_lc_update()

**Review by kccain — COMMENTED** [review]
did a first pass, have only a couple of questions. Looks good overall. I may need to re-read it again to make sure I fully absorb it. It sounds like this patch is going to be waiting for some vos implementation to land first, so there should be some more time for that?

File: src/rdb/rdb_tx.c
Comment (liw) [reply src/rdb/rdb_tx.c:222]: A decoder here, such as `rdb_tx_hdr_decode`, takes a sequence of bytes, and outputs how many bytes have been consumed as well as the decoded data. The number of consumed bytes may or may not be equal to the size of the decoded data, although they are indeed equal in this case at the moment. For instance, if we were crazy enough to encode in JSON... Hence, I think it's better to avoid using the internal knowledge (i.e., the number of bytes consumed) outside the decoder in non-test code.

**Comment by liw** [comment]
Rebased without changes to the update janekmi/local_tx2 branch.

**Comment by liw** [comment]
daos_test/suite.py: drain_simple: Wolf IB network problem

**Comment by liw** [comment]
Rebased back to master.

**Comment by liw** [comment]
Rebased to liw/fix-dth_deferred so that testing can continue.

**Comment by liw** [comment]
Per request, starting the review process before getting the latest CI result.

File: src/rdb/rdb_raft.c
Comment (NiuYawei) [inline src/rdb/rdb_raft.c:1282]: Do you know what's the typical transaction size after this change?

FYI, there is a potential bug in PMDK which could be triggered when the transaction redo log is large. (see DAOS-13081 and https://github.com/pmem/pmdk/issues/5583), but seems we never hit the issue so far, I guess it because the redo log size (not including snapshot which is in undo log) is usually very small for PMDK transaction.

**Review by NiuYawei — DISMISSED** [review]

**Comment by liw** [comment]
Rebased due to the base branch change. No code changes.

**Comment by liw** [comment]
daos_test/rebuild_simple: DAOS-15290

**Review by NiuYawei — DISMISSED** [review]

**Review by kccain — DISMISSED** [review]

**Review by kccain — APPROVED** [review]

**Comment by liw** [comment]
rebuild_simple REBUILD27: DAOS-15647

**Review by NiuYawei — APPROVED** [review]

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**
